xclim/sdba/_adjustment.py

```diff
@@ -126,32 +141,315 @@ def eqm_train(
+    Do this iteratively over all rotations `rots` and conserve adjustment factors `af_q` in each iteration.
+    """
+    if standardize:
+        ref = (ref - np.nanmean(ref, axis=-1, keepdims=True)) / (
+            np.nanstd(ref, axis=-1, keepdims=True)
+        )
+        hist = (hist - np.nanmean(hist, axis=-1, keepdims=True)) / (
+            np.nanstd(hist, axis=-1, keepdims=True)
+        )
+    af_q = np.zeros((len(rots), ref.shape[0], len(quantiles)))
+    escores = np.zeros(len(rots)) * np.NaN
+    if n_escore > 0:
+        ref_step, hist_step = (
+            int(np.ceil(arr.shape[1] / n_escore)) for arr in [ref, hist]
+        )
+    for ii in range(len(rots)):
+        rot = rots[0] if ii == 0 else rots[ii] @ rots[ii - 1].T
+        ref, hist = (rot @ da for da in [ref, hist])
+        # loop over variables
+        for iv in range(ref.shape[0]):
+            ref_q, hist_q = (nbu._quantile(da, quantiles) for da in [ref[iv], hist[iv]])
+            af_q[ii, iv] = u.get_correction(hist_q, ref_q, "+")
+            af = u._interp_on_quantiles_1D(
+                u._rank_bn(hist[iv]),
+                quantiles,
+                af_q[ii, iv],
+                method=method,
+                extrap=extrap,
+            )
+            hist[iv] = u.apply_correction(hist[iv], af, "+")
+        if n_escore > 0:
+            escores[ii] = nbu._escore(ref[:, ::ref_step], hist[:, ::hist_step])
+    hist = rots[-1].T @ hist
+    return af_q, escores
+
+
+def mbcn_train(
+    ds,
+    rot_matrices,
+    pts_dims,
+    quantiles,
+    g_idxs,
+    gw_idxs,
+    interp,
+    extrapolation,
+    n_escore,
+) -> xr.Dataset:
+    """Npdf transform training.
+
+    Adjusting factors obtained for each rotation in the npdf transform and conserved to be applied in
+    the adjusting step in ``mcbn_adjust``.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset variables:
+            ref : training target  (stacked variables)
+            hist : training source (stacked variables)
+    rot_matrices : xr.DataArray
+        The rotation matrices as a 3D array ('iterations', <pts_dims[0]>, <pts_dims[1]>), with shape (n_iter, <N>, <N>).
+    pts_dims : str
+        The name of the "multivariate" dimension and its primed counterpart. Defaults to "multivar", which
+        is the normal case when using :py:func:`xclim.sdba.base.stack_variables`, and "multivar_prime",
+    g_idxs : xr.DataArray
+        Indices of the times in each time group
+    gw_idxs : xr.DataArray
+        Indices of the times in each windowed time group
+    """
+    # unpack data
+    ref = ds.ref
+    hist = ds.hist
+    gr_dim = gw_idxs.attrs["group_dim"]
+
+    # npdf training core
+    af_q_l = []
+    escores_l = []
+
+    # loop over time blocks
+    for ib in range(gw_idxs[gr_dim].size):
+        # indices in a given time block
+        indices = gw_idxs[{gr_dim: ib}].astype(int).values
+        ind = indices[indices >= 0]
```

Comment:
¿ Por qué ?

Reply:
¿ que lo qué?

Reply:
That's probably related to memory leaks. I guess it's not the most xarray-esque way of doing things, I have to improve this.

Reply:
I meant the `indices >= 0`, I don't think I understand.

Reply:
it removes possible nans (which can come up with standard calendar because of leap years iirc, but I would not recommend using this with a non-regular calendar anyways)

Reply:
So you are relying on that fact that NaN translates to a negative number with `astype` ? If so I suggest:
```python3
        indices = gw_idxs[{gr_dim: ib}].fillna(-1).astype(int).values
        ind = indices[indices >= 0]
```